Sequence Parallel

ZYHowell opened the pull request:
## Motivation

When serving an extremely large model (e.g. Llama 400B), the #GPU might be more than #kv head. This leads to a replication on kv cache, which is troublesome when the sequence length is too large

## Modification

This PR introduced a very basic sequence parallelism on the attention computation. For all other parts, the model is still fully tensor parallelized. The partition switches before and after the attention. This is achieved by:
1. When preparing the batch, collocate input ids on the same sequence parallel rank (`sp_rank`) together, this is referred as the [sequence parallel layout](https://github.com/sgl-project/sglang/pull/1041/files#diff-79875277a348b67c6404172a31215ebf379de9c6f88f46b66c5ab3639ba82bfcR336-R345) in this pr and the code comments. Flash infer args are accordingly changed;
2. Before entering the SP part, only the KV locally stored is computed. (python/sglang/srt/layers/linear.py)
3. The SP kernel, which still has some space to improve. (python/sglang/srt/layers/radix_attention.py)
4. When leaving the SP part, the whole sequence is collected again, because the rest part takes the whole sequence.
5. The output logits are switched back before doing the sampling.
6. MISC modification including: Parallel State (python/sglang/srt/layers/parallel_utils/parallel_state.py), calling all components in model runner (python/sglang/srt/managers/controller/model_runner.py) and the model definition (python/sglang/srt/models/llama2.py), server args (python/sglang/srt/server_args.py), and tests

## Checklist

- [ ]  Ensure pre-commit `pre-commit run --all-files` or other linting tools are used to fix potential lint issues.
- [ ] Confirm that modifications are covered by complete unit tests. If not, please add more unit tests for correctness.
- [ ]  Modify documentation as needed, such as docstrings or example tutorials.

ZYHowell committed:
add sp index

ZYHowell committed:
add clone for rope as it's in-place

ZYHowell committed:
add decode mask for sp

ZYHowell committed:
insert to prepare batch

ZYHowell committed:
add sp size and rank args

ZYHowell committed:
update sequence parallel layout

ZYHowell committed:
minor bug fix to pass sp=1 test

ZYHowell committed:
give local indices to help with position ids; prepare for only record local kv cache

ZYHowell committed:
minor fix

ZYHowell committed:
add sp layout to normal layout

ZYHowell committed:
move sp layout transform tool to inputmetadata

ZYHowell committed:
add debug flatten to sp

ZYHowell committed:
update name and doc string

ZYHowell committed:
bug fix for the sp=1 case

ZYHowell committed:
fix prefix lens None

ZYHowell committed:
fix debug mode indices

ZYHowell committed:
runnable but only first two decode tok cor

ZYHowell committed:
fix early exit for decode with SP

ZYHowell committed:
format

Commit:
Merge pull request #1 from ivanium/pr-sp-rope

Sequence Parallel system setup

ZYHowell committed:
Update sp layout (#3)

* update layout

* bug fix

Commit:
Sequence parallel prefill attention kernel (#2)

* test: test cases of combining multiple attention kernel calls to implement a sequence parallel kernel. Verified with 2 sp workers

* fix: simplify flashinfer kernel initialization (begin_forward() and end_forward())

* test: add logic for sp worker 1 which is basically the same but with different orders of kernel calls

* chore: format tweak

* feat: a general seq parallel attention kernel that achieves workload balance

* fix: minor tweak loop iteration within ring attention

* feat [radix_attention]: seq_parallel kernel with sync communication.

TODO: turn communication into async fashion and overlap it with computation

* test: update test cases for seq parallel attn kernel. Need to disable kv cache management before testing because we haven't implemented kv cache management for seq parallel yet

* chore [radix_attention]: format tweak

* feat: async communication within ring attention

* fix [parallel_utils]: add missed files

* fix [infer_batch]: set default values for newly added sp-related metadata

* fix [bench_latency]: minor fixes to input args

* feat [parallel_utils]: get actual tp rank and size when both TP and SP are enabled

* feat [linear]: add QKVParallelLinear

* feat [llama2]: update llama model to use our QKVParallelLinear

* feat [model_runner]: initialize model parallel with sequence parallel

* fix [infer_batch]: 1. a minor issue when calling get_prefill_indices; 2. flashinfer intialization args

* fix [bench_latency]: load model with sp_rank

* feat [radix_attention]: automatically dispatch to seq-parallel attn kernel when sp_size > 1

* debug: stash current debug changes

* fix [radix_attention]: reshape q tensor before running the kernel

* bug fix for sp layout types

* fix: adjust tensor layout. TODO: fix many dirty hacks and hardcoded values

* fix [wip]: disable p2p communication within ring attention for now. TODO: fix the bug that causes communication hang.

* chore [bench_latency]: disable decode for now since we haven't supported it

* upstream with correct prefill sp layout

* fix early exit on decode SP

* chore: tweak format

* update layout

* bug fix

* fix [linear, radix_attention]: fix q head indexes per SP worker to align with GQA setting.

* fix [infer_batch]: set up flashinfer kernels for the batch size > 1 case

* chore: tweak format

* fix [radix_attention]: revert commented-out kv cache store operations in normal attention

* fix: adjust k, v tensor shape to align with both TP and SP setting

* chore [llama2]: minor adjustment

* fix: update bench_latency to evenly distribute each sequence across all SP workers to avoid the layout issue

* test: update test cases to align with current kernel in args

* fix [model_runner]: initialize TokenToKVPool with correct num_heads and enable KV cache store in SP attention

* chore [radix_attention]: clean up comments

* fix [model_runner]: correct num_heads in memory profiling as well to avoid OOM

* fix [infer_batch]: adopt SP KV cache allocation

* feat [linear]: correctly partition q proj along the num_heads dimension with GQA

* chore [llama2]: clean up stable variables

* feat [infer_batch]: adjust positions to SP layout when preparing input_metadata

* feat [infer_batch]: use dedicate paged attn kernel for cross-SP-shard attn

* feat [parallel_state]: creat sequence parallel comm groups

* test [sp_comm_group]: simple test case with sp_size = 2

* doc [parallel_state]: doc string for our SP group organization

* fix [infer_batch]: add padding zeros to positions tensor and out_cache_loc to fix positional encoding and KV cache store

* feat [radix_attn, infer_batch]: create masks for padded sequences and now attn works for unevenly-distributed sequenses too

* chore [bench_latency]: revert original prompts

* fix [parallel_state]: rename "actual" to "kv"

* refactor [radix_attention]: unified two cases with differnt comm-comp tradeoffs

* chore: rename "actual_tp_[size|rank]" to "kv_tp_[size|rank]"

* fix [infer_batch]: ensure prefix_lens is not None in init_flashinfer_args

* fix [infer_batch]: only pad positions and out_cache_loc for prefill

* chore [linear]: clean up and revise comments

* chore [parallel_state]: revise comments

* chore [linear]: revise comments and class names

* chore [radix_attention]: add defensive checks

---------

Co-authored-by: ZYHowell <yhzhuang@cmu.edu>

Commit:
Sequence Parallel Decode Attn Kernel (#5)

Commit:
fix [infer_batch]: fix _get_decode_local_lens and use it to initialize flashinfer decode kernel (#6)

zhyncs commented:
@ZYHowell Thanks for your contribution! May you rebase the latest main branch and resolve the conflicts? Thanks!

ZYHowell committed:
all changes

ZYHowell committed:
merge

ZYHowell committed:
Merge remote-tracking branch 'origin/main' into pr-rebase

ZYHowell committed:
fix lint and test import

ivanium committed:
fix [sp_linear, llama2]: add prefix support for SP linear layers. Fix llama weight load

ivanium committed:
fix [forward_batch_info]: prefix_lens can be None in decode phase so avoid it in update_flashinfer_indices

ivanium committed:
fix [radix_attention]: remove duplicated sp extend/decode functions

ivanium committed:
fix [radix_attention]: kv_data -> get_kv_buffer()

ZYHowell committed:
minor bug fix

ZYHowell committed:
fix out cache loc

ivanium committed:
fix [forward_batch_info]: init flashinfer ragged kernel correctly for SP

ivanium committed:
fix [cuda_graph_runner]: add missing SP parameters to CUDA graph initialization. TODO: init SP parameters correctly

ZYHowell committed:
Merge pull request #9 from ivanium/pr-rebase

ZYHowell committed:
Merge remote-tracking branch 'upstream/main' into pr-rebase-2561ed0

ZYHowell committed:
minor fix

ZYHowell committed:
contiguous and to device for triton kernel

ZYHowell committed:
fix paged kernel lens before kv indptr

Commit:
Merge pull request #10 from ivanium/pr-rebase-2561ed0

rebase to 2561ed0

ivanium committed:
Sequence parallel prefill attention kernel (#2)

* test: test cases of combining multiple attention kernel calls to implement a sequence parallel kernel. Verified with 2 sp workers

* fix: simplify flashinfer kernel initialization (begin_forward() and end_forward())

* test: add logic for sp worker 1 which is basically the same but with different orders of kernel calls

* chore: format tweak

* feat: a general seq parallel attention kernel that achieves workload balance

* fix: minor tweak loop iteration within ring attention

* feat [radix_attention]: seq_parallel kernel with sync communication.

TODO: turn communication into async fashion and overlap it with computation

* test: update test cases for seq parallel attn kernel. Need to disable kv cache management before testing because we haven't implemented kv cache management for seq parallel yet

* chore [radix_attention]: format tweak

* feat: async communication within ring attention

* fix [parallel_utils]: add missed files

* fix [infer_batch]: set default values for newly added sp-related metadata

* fix [bench_latency]: minor fixes to input args

* feat [parallel_utils]: get actual tp rank and size when both TP and SP are enabled

* feat [linear]: add QKVParallelLinear

* feat [llama2]: update llama model to use our QKVParallelLinear

* feat [model_runner]: initialize model parallel with sequence parallel

* fix [infer_batch]: 1. a minor issue when calling get_prefill_indices; 2. flashinfer intialization args

* fix [bench_latency]: load model with sp_rank

* feat [radix_attention]: automatically dispatch to seq-parallel attn kernel when sp_size > 1

* debug: stash current debug changes

* fix [radix_attention]: reshape q tensor before running the kernel

* bug fix for sp layout types

* fix: adjust tensor layout. TODO: fix many dirty hacks and hardcoded values

* fix [wip]: disable p2p communication within ring attention for now. TODO: fix the bug that causes communication hang.

* chore [bench_latency]: disable decode for now since we haven't supported it

* upstream with correct prefill sp layout

* fix early exit on decode SP

* chore: tweak format

* update layout

* bug fix

* fix [linear, radix_attention]: fix q head indexes per SP worker to align with GQA setting.

* fix [infer_batch]: set up flashinfer kernels for the batch size > 1 case

* chore: tweak format

* fix [radix_attention]: revert commented-out kv cache store operations in normal attention

* fix: adjust k, v tensor shape to align with both TP and SP setting

* chore [llama2]: minor adjustment

* fix: update bench_latency to evenly distribute each sequence across all SP workers to avoid the layout issue

* test: update test cases to align with current kernel in args

* fix [model_runner]: initialize TokenToKVPool with correct num_heads and enable KV cache store in SP attention

* chore [radix_attention]: clean up comments

* fix [model_runner]: correct num_heads in memory profiling as well to avoid OOM

* fix [infer_batch]: adopt SP KV cache allocation

* feat [linear]: correctly partition q proj along the num_heads dimension with GQA

* chore [llama2]: clean up stable variables

* feat [infer_batch]: adjust positions to SP layout when preparing input_metadata

* feat [infer_batch]: use dedicate paged attn kernel for cross-SP-shard attn

* feat [parallel_state]: creat sequence parallel comm groups

* test [sp_comm_group]: simple test case with sp_size = 2

* doc [parallel_state]: doc string for our SP group organization

* fix [infer_batch]: add padding zeros to positions tensor and out_cache_loc to fix positional encoding and KV cache store

* feat [radix_attn, infer_batch]: create masks for padded sequences and now attn works for unevenly-distributed sequenses too

* chore [bench_latency]: revert original prompts

* fix [parallel_state]: rename "actual" to "kv"

* refactor [radix_attention]: unified two cases with differnt comm-comp tradeoffs

* chore: rename "actual_tp_[size|rank]" to "kv_tp_[size|rank]"

* fix [infer_batch]: ensure prefix_lens is not None in init_flashinfer_args

* fix [infer_batch]: only pad positions and out_cache_loc for prefill

* chore [linear]: clean up and revise comments

* chore [parallel_state]: revise comments

* chore [linear]: revise comments and class names

* chore [radix_attention]: add defensive checks

---------

Co-authored-by: ZYHowell <yhzhuang@cmu.edu>

ZYHowell committed:
Merge pull request #11 from ivanium/pr-rebase-ab4a83b

Rebase ab4a83b

merrymercy commented:
moved to #1436
